framework/TSA/RWD.py

```diff
@@ -0,0 +1,244 @@
+    settings['FeatureIndex'] = spec.findFirst('FeatureIndex').value
+    settings['SampleType'] = spec.findFirst('SampleType').value
+
+
+    return settings
+
+  def setDefaults(self, settings):
+    """
+      Fills default values for settings with default values.
+      @ In, settings, dict, existing settings
+      @ Out, settings, dict, modified settings
+    """
+    settings = super().setDefaults(settings)
+    if 'SignatureWindowLength' not in settings:
+      settings['SignatureWindowLength'] = None #len(history)//10
+      settings['SampleType'] = 1
+
+    return settings ####
+
+  def characterize(self, signal, pivot, targets, settings):
+    """
+      Determines the charactistics of the signal based on this algorithm.
+      @ In, signal, np.ndarray, time series with dims [time, target]
+      @ In, pivot, np.1darray, time-like parameter values
+      @ In, targets, list(str), names of targets in same order as signal
+      @ In, settings, dict, settings for this ROM
+      @ Out, params, dict of dict: 1st level contains targets/variables; 2nd contains: U vectors and features
+    """
+    # lazy import statsmodels
+    import statsmodels.api
+    # settings:
+    #   SignatureWindowLength, int,  Signature window length
+    #   FeatureIndex, list of int,  The index that contains differentiable params
+
+
+    params = {}
+
+    for tg, target in enumerate(targets):
+      history = signal[:, tg]
+      if settings['SignatureWindowLength'] is None:
+        settings['SignatureWindowLength'] = len(history)//10
+      SignatureWindowLength = int(settings['SignatureWindowLength'])
+      fi = int(settings['FeatureIndex'])
+      SampleType = settings['SampleType']
+      AllWindowNumber = int(len(history)-SignatureWindowLength+1)
+      SignatureMatrix = np.zeros((SignatureWindowLength, AllWindowNumber))
+      for i in range(AllWindowNumber):
+        SignatureMatrix[:,i] = np.copy(history[i:i+SignatureWindowLength])
+
+      # Sequential sampling
+      if SampleType == 0:
+        BaseMatrix = np.copy(SignatureMatrix)
+
+      # Randomized sampling
+      elif SampleType == 1:
+        sampleLimit = len(history)-SignatureWindowLength
+
+        WindowNumber = sampleLimit//4
+        sampleIndex = np.random.randint(sampleLimit, size=WindowNumber)
+        BaseMatrix = np.zeros((SignatureWindowLength, WindowNumber))
+        for i in range(WindowNumber):
+          WindowIndex = sampleIndex[i]
+          BaseMatrix[:,i] = np.copy(history[WindowIndex:WindowIndex+SignatureWindowLength])
+      # Piecewise Sampling
+      else:
+        WindowNumber = len(history)//SignatureWindowLength
+        BaseMatrix = np.zeros((SignatureWindowLength, WindowNumber))
+        for i in range(WindowNumber-1):
+          BaseMatrix[:,i] = np.copy(history[i*SignatureWindowLength:(i+1)*SignatureWindowLength])
+      U,s,V = mathUtils.computeTruncatedSingularValueDecomposition(BaseMatrix,0)
+      FeatureMatrix = U.T @ SignatureMatrix
```

Comment:
Please use camelBack style for 'SampleType', 'AllWindowNumber', 'BaseMatrix', ...